### PR TITLE
ALTAPPS-679: Tracks, progresses & studyPlan cached access

### PR DESCRIPTION
**Pull request by XanderZhu**
**YouTrack Issues**:
[#ALTAPPS-679](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-679)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
`RepositoryCacheProxy` - is a proxy that should be used to fetch values by keys from the cache or to load them from the network. Its behavior is similar to BaseStateRepository. It takes `InMemoryRepositoryCache` by default.
It is implemented as a final class, not an abstract because it could be instantiated many times for different entities inside the repository (see [ProgressRepositorImpl](https://github.com/hyperskill/mobile-app/blob/a49a6193367e12bd9f7217b86028049a1c09daef/shared/src/commonMain/kotlin/org/hyperskill/app/progresses/data/repository/ProgressesRepositoryImpl.kt)).
In short, `RepositoryCacheProxy` detects keys, which should be loaded from the network, writes their values into the cache, and mixes remote values with cached.

This approach is supposed to be used for `Tracks` in the progress screen, progress in `GamificationToolbarFeature`, and `StudyPlanWidgetFeature`.

### CODE REVIEW

**Comment by XanderZhu** [comment]
@vladkash @ivan-magda 
At the moment this branch contains a cache only for `ProgressRepository` as an example.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/InMemoryRepositoryCache.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.hyperskill.app.core.data.repository_cache
+
+class InMemoryRepositoryCache<Key : Any, Value : Any?> : RepositoryCache<Key, Value> {
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/InMemoryRepositoryCache.kt:3]: Key - is a generic parameter.
Any - is a low constraint for Key to not allow Nothing as a Key.
When u create an instance of Cache with your consttaints.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/SingletonRepositoriesComponentImpl.kt [hunk 2]
```diff
@@ -1,17 +1,29 @@
...
+        SubscriptionsRemoteDataSourceImpl(authorizedHttpClient)
 
     override val currentSubscriptionStateRepository: CurrentSubscriptionStateRepository =
         CurrentSubscriptionStateRepositoryImpl(subscriptionsRemoteDataSource)
+
+    /**
+     * Progress
+     */
+    override val progressRepository: ProgressesRepository
+        get() = ProgressesRepositoryImpl(ProgressesRemoteDataSourceImpl(authorizedHttpClient))
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/SingletonRepositoriesComponentImpl.kt:28]: Ooops, my fault

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/InMemoryRepositoryCache.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/InMemoryRepositoryCache.kt:3]: All classes has hashCode, jus because they are subclasses of Any

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt:1]: > some repository uses cache proxy it must be singleton so that it can keep the cache alive and share it between other users (features)

We can extract the cache to the repository constructor. But anyway, in the case of in-memory cache, we have to store it as a singleton

[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt:1]: @ivan-magda 
I've just implemented `TrackProgressesCacheDataSource` & `ProjectProgressesCacheDataSource` as we usually do.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    keysToLoadFromRemote.add(key)
+                }
+            }
+
+            if (keysToLoadFromRemote.isEmpty()) {
+                Result.success(cachedValues)
+            } else {
+                loadManyValuesInternal(keysToLoadFromRemote)
+                    .map { remoteValues ->
+                        cachedValues + remoteValues
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt:60]: @ivan-magda 
What is the reason to sort results by key?
We don't know which order is preferred for the user of the `RepositoryCacheProxy`. In my opinion, it is a user's responsibility to sort the results they want.

**Comment by XanderZhu** [comment]
@ivan-magda 
I realized that the current `ReposiitoryCache` interface can not be implemented by DB cache: it would be not optimized to make many small DB requests by id. This is why I changed `ReposiitoryCache` interface.

File: shared/src/commonMain/kotlin/org/hyperskill/app/placeholder_new_user/presentation/PlaceholderNewUserActionDispatcher.kt [hunk 4]
```diff
@@ -35,7 +35,7 @@ class PlaceholderNewUserActionDispatcher(
                     .filter { it.isBeta.not() }
 
                 val trackProgressById = progressesInteractor
-                    .getTracksProgresses(tracks.map { it.id })
+                    .getTracksProgresses(tracks.map { it.id }, forceLoadFromRemote = action.forceUpdate)
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/placeholder_new_user/presentation/PlaceholderNewUserActionDispatcher.kt:38]: Maybe it isn't necessary to add force update logic here. Initialize message is send only when screen is opened first time and after network error reloading. In both cases we should load from remote, so we can use `true` here, without forceUpdate param

File: shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/interactor/ProgressesInteractor.kt [hunk 5]
```diff
@@ -8,21 +8,21 @@ import org.hyperskill.app.track.domain.model.TrackProgress
...
 
-    suspend fun getProjectsProgresses(projectsIds: List<Long>): Result<List<ProjectProgress>> =
-        progressesRepository.getProjectsProgresses(projectsIds)
+    suspend fun getProjectsProgresses(projectsIds: List<Long>, forceLoadFromRemote: Boolean): Result<List<ProjectProgress>> =
+        progressesRepository.getProjectsProgresses(projectsIds, forceLoadFromRemote)
 
-    suspend fun getProjectProgress(projectId: Long): Result<ProjectProgress> =
-        progressesRepository.getProjectProgress(projectId)
+    suspend fun getProjectProgress(projectId: Long, forceLoadFromRemote: Boolean): Result<ProjectProgress> =
+        progressesRepository.getProjectProgress(projectId, forceLoadFromRemote)
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/interactor/ProgressesInteractor.kt:27]: if we don't have any business logic and just delegate in interactor, why can't we use progressesRepository directly?

File: shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/repository/ProgressesRepository.kt [hunk 6]
```diff
@@ -17,10 +17,12 @@ interface ProgressesRepository {
...
+    suspend fun getProjectsProgresses(projectsIds: List<Long>, forceLoadFromRemote: Boolean): Result<List<ProjectProgress>>
 
-    suspend fun getProjectProgress(projectId: Long): Result<ProjectProgress> =
+    suspend fun getProjectProgress(projectId: Long, forceLoadFromRemote: Boolean): Result<ProjectProgress> =
         kotlin.runCatching {
-            getProjectsProgresses(listOf(projectId)).getOrThrow().first()
+            getProjectsProgresses(listOf(projectId), forceLoadFromRemote).getOrThrow().first()
         }
+
+    fun clearCache()
```
Comment (vladkash) [inline shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/repository/ProgressesRepository.kt:27]: This function isn't used, we're clearing cache after deauthorization in `SingletonRepositoriesComponent`

**Review by vladkash — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/interactor/ProgressesInteractor.kt (hunk 5 above)
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/progresses/domain/interactor/ProgressesInteractor.kt:27]: I agree. I don't know the reason we have `ProgressInteractor`. The major of its methods is not used.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/data/repository_cache/RepositoryCacheProxy.kt]: Because the backend does this - when objects are requested by ids, the result objects will be sorted by the originally requested ids.

We may already rely on that backend behavior when implementing some features. Cache proxy will return result objects in unpredictable order, so that may lead to some bugs.

If we will not add sorting to the cache proxy, we must check current usages and remember in the future that result objects will not be sorted by ids.

**Comment by ivan-magda** [comment]
> I realized that the current `RepositoryCache` interface can not be implemented by DB cache: it would be not optimized to make many small DB requests by id. This is why I changed `ReposiitoryCache` interface.

@XanderZhu when we will add some DB, we want to perform queries on that, not return all available objects and do this in memory...

File: shared/src/commonMain/kotlin/org/hyperskill/app/track/domain/model/TrackProgress.kt [hunk 7]
```diff
@@ -1,13 +1,15 @@
 package org.hyperskill.app.track.domain.model
 
-import kotlin.math.roundToInt
 import kotlinx.serialization.SerialName
 import kotlinx.serialization.Serializable
+import kotlin.math.roundToInt
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/track/domain/model/TrackProgress.kt:5]: Why when I rearrange imports `kotlin.math.roundToInt` moves up 🤔 

My settings:

Screenshot 2023-04-06 at 12 59 40 PM

**Review by ivan-magda — APPROVED** [review]
@XanderZhu please check one more time what repositories/caches were created/modified and don't forget to call `clearCache` for them.

Also, please take a look at this "[comment](https://github.com/hyperskill/mobile-app/pull/423#discussion_r1159267721)".

File: shared/src/commonMain/kotlin/org/hyperskill/app/progresses/injection/ProgressesDataComponentImpl.kt [hunk 8]
```diff
@@ -1,21 +1,40 @@
...
-    override val progressesInteractor: ProgressesInteractor
-        get() = ProgressesInteractor(progressesRepository)
+    companion object {
+        private val trackProgressesCacheDataSource: TrackProgressesCacheDataSource by lazy {
+            TrackProgressesCacheDataSourceImpl(InMemoryRepositoryCache())
+        }
+        private val projectProgressesCacheDataSource: ProjectProgressesCacheDataSource by lazy {
+            ProjectProgressesCacheDataSourceImpl(InMemoryRepositoryCache())
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/progresses/injection/ProgressesDataComponentImpl.kt:39]: and in all other places

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/StateRepositoriesComponentImpl.kt [hunk 9]
```diff
@@ -1,17 +1,30 @@
...
-        SubscriptionsRemoteDataSourceImpl(appGraph.networkComponent.authorizedHttpClient)
+        SubscriptionsRemoteDataSourceImpl(authorizedHttpClient)
 
     override val currentSubscriptionStateRepository: CurrentSubscriptionStateRepository =
         CurrentSubscriptionStateRepositoryImpl(subscriptionsRemoteDataSource)
+
+    /**
+     * Study plan
+     */
+    override val studyPlanStateRepository: CurrentStudyPlanStateRepository by lazy {
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/StateRepositoriesComponentImpl.kt:27]: @ivan-magda 
Because we don't need it until someone requests it.
On the first call repository will be created, and other requestors will get cached value.

File: shared/src/commonMain/kotlin/org/hyperskill/app/main/domain/interactor/AppInteractor.kt [hunk 10]
```diff
@@ -3,22 +3,29 @@ package org.hyperskill.app.main.domain.interactor
...
     private val analyticInteractor: AnalyticInteractor,
+    private val progressesRepository: ProgressesRepository,
+    private val trackRepository: TrackRepository
 ) {
     suspend fun doCurrentUserSignedOutCleanUp() {
         analyticInteractor.flushEvents()
         clearCache()
     }
 
+    // TODO: define more flexible & scalable way to clear cache
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/main/domain/interactor/AppInteractor.kt:23]: At least we can pass a List<Clearable> or smth else, to not modify AppInteractor every time a new cache appears.

File: shared/src/commonMain/kotlin/org/hyperskill/app/progresses/injection/ProgressesDataComponentImpl.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/progresses/injection/ProgressesDataComponentImpl.kt:39]: Ok, I've just used standard kotlin style guide
https://kotlinlang.org/docs/coding-conventions.html#class-layout

File: shared/src/commonMain/kotlin/org/hyperskill/app/track/data/repository/TrackRepositoryImpl.kt [hunk 11]
```diff
@@ -1,15 +1,33 @@
...
-        trackRemoteDataSource.getTracks(trackIds)
+
+    private val trackCacheProxy = RepositoryCacheProxy(
+        cache = trackCacheDataSource,
+        loadValuesFromRemote = { trackIds ->
+            trackRemoteDataSource.getTracks(trackIds)
+        },
+        getKeyFromValue = { track ->
+            track.id
+        }
```
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/track/data/repository/TrackRepositoryImpl.kt:21]: The alternative is to replace `track` with `it` so that you don't need to name the lambda parameter.
I decided to name parameters in this case for better readability.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/StateRepositoriesComponentImpl.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (XanderZhu) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/injection/StateRepositoriesComponentImpl.kt:27]: @ivan-magda 
Di graph is initialized just after the app startup. We need studyPlanStateRepository quite later.

File: shared/src/commonMain/kotlin/org/hyperskill/app/core/domain/repository_cache/RepositoryCacheProxy.kt [hunk 12]
```diff
@@ -36,7 +36,7 @@ class RepositoryCacheProxy<in Key : Any, Value : Any?>(
                 } else {
                     loadValuesAndUpdateCache(keysToLoadFromRemote.toList())
                         .map { remoteValues ->
-                            cachedValues + remoteValues
+                            (cachedValues + remoteValues).sortedBy(getKeyFromValue)
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/core/domain/repository_cache/RepositoryCacheProxy.kt:39]: @XanderZhu I mean sorting by parameter `keys: List<Key>`, `getValues(keys = [2, 5, 1, 7])` -> result list sorted in order of `[2, 5, 1, 7]`

For example [stepik-ios implementation](https://github.com/StepicOrg/stepik-ios/blob/master/Stepic/Sources/Extensions/Foundation/Array%2BReordering.swift)

Comment (ivan-magda) [reply shared/src/commonMain/kotlin/org/hyperskill/app/core/domain/repository_cache/RepositoryCacheProxy.kt:39]: I'll ask the backend team about do they sorting for every endpoint when getting objects by ids or not. If not then we don't need sorting here.
